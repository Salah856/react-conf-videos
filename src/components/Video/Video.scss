.root {
  width: 100%;
  padding: 1rem 2rem;
  box-sizing: border-box;
  font-size: 1.6rem;

  .heading {
    margin: 0;
  }

  .top {
    display: flex;
    justify-content: space-between;
    width: 100%;
    cursor: pointer;

    background: transparent;
    border: none;
    padding: 0;

    font-size: 16px;

    .title {
      font-weight: bold;
      text-align: left;
    }

    .right {
      font-size: 1.6rem;
      padding-left: 0.5rem;
    }
  }

  .videoWrapper {
    height: 0;
    width: 100%;
    overflow-y: hidden;
    transition: height 0.5s;
    background: url('../../../public/assets/tv.svg');
    background-repeat: no-repeat;
    background-size: 150px 150px;
    background-position: center center;

    &.open {
      height: 360px;
      margin: 1rem 0;
      background-color: black;
    }
  }

  .details {
    display: flex;
    justify-content: space-between;
    font-size: 1.4rem;

    .conferenceTitle {
      cursor: pointer;
      white-space: nowrap;
    }
  }

  .divider {
    height: 1px;
    width: 100%;
    background-color: grey;
  }
}
