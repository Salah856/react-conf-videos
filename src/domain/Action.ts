export type Action<T> = {
  type: string
  payload?: T
  error?: boolean
  meta?: object
}
