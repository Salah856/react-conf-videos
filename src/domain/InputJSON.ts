/* Definitions for input JSON file read in */
export type PresenterInput = {
  name: string;
};

export type VideoInput = {
  title: string;
  link: string;
  length: string;
  split: string;
  presenter: PresenterInput;
};

export type ConferenceInput = {
  title: string;
  date: string;
  website: string;
  playlist: string;
  videos: VideoInput[];
};

export type JSONInput = ConferenceInput[];
