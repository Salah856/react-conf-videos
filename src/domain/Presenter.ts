export type Presenter = {
  name: string;
};

export type IndexedPresenters = {
  [idx: string]: Presenter;
};
