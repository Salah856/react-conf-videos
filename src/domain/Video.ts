export type Video = {
  link: string
  title: string
  split: string
  length: string
  lightning?: boolean
  presenter: string // id
  embeddableLink: string
}

export type IndexedVideos = {
  [idx: string]: Video
}
