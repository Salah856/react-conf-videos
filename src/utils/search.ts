import { compose, toLower, trim } from 'ramda'
import { remove as removeDiacritics } from 'diacritics'

export const cleanQuery = compose(
  removeDiacritics,
  toLower,
  trim
)

export const isFilterEmpty = (filterValue: string = '') => filterValue.trim() === ''
