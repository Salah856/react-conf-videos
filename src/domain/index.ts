import { Action } from './Action';
import {
  Conference,
  ConferenceTitlesToIds,
  IndexedConferences
} from './Conference';
import {
  ConferenceInput,
  JSONInput,
  PresenterInput,
  VideoInput
} from './InputJSON';
import { IndexedPresenters, Presenter } from './Presenter';
import { IndexedVideos, Video } from './Video';

export {
  Action,
  Video,
  IndexedVideos,
  Presenter,
  IndexedPresenters,
  Conference,
  IndexedConferences,
  ConferenceTitlesToIds,
  ConferenceInput,
  VideoInput,
  PresenterInput,
  JSONInput
};
