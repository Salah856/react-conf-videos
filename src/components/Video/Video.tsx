import * as React from 'react'
import * as cn from 'classnames'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { compose, pure, withStateHandlers } from 'recompose'

import { Video as VideoProp, Presenter, Conference } from '../../domain'
import { ApplicationState } from 'redux/modules'
import { sluggifyUrl } from 'utils'

import styles from './Video.scss'

type Props = {
  videoId: string
  conferenceId: string
}

type State = {
  isOpen: boolean
}

type StateHandlers = {
  toggleIsOpen: (e: React.MouseEvent<HTMLElement>) => State
}

type ReduxState = {
  conferenceId: string
  conference: Conference
  speaker: Presenter
  video: VideoProp
}

type CombinedProps = Props & State & ReduxState & StateHandlers

export const VideoInner: React.SFC<CombinedProps> = ({
  video: { title, length, embeddableLink },
  speaker, videoId, conference, isOpen, toggleIsOpen, conferenceId
}) => {
  return (
    <div className={styles.root} key={videoId} >
      <div className={styles.top} onClick={toggleIsOpen} >
        <span className={styles.title}>{title}</span>
        <span className={styles.right}>{length}</span>
      </div>
      <div className={cn(styles.videoWrapper, { [styles.open]: isOpen })}>
        {isOpen && <iframe
          title="videoPlayer"
          id="ytplayer"
          width="410"
          height="360"
          src={embeddableLink}
          frameBorder="0"
          allowFullScreen={true}
        /> }
      </div>
      <div className={styles.details}>
        <span>{speaker.name}</span>
        <Link to={`/conference/${conferenceId}/${sluggifyUrl(conference.title)}`}>
          <span className={styles.conferenceTitle}>{conference.title}</span>
        </Link>
      </div>
    </div>
  )
}

const mapStateToProps = (state: ApplicationState, props: Props) => {
  const { data: { videos, presenters, conferences } } = state
  const { videoId, conferenceId } = props
  const video = videos[videoId]
  const conference = conferences[conferenceId]
  const speaker = presenters[video.presenter]
  return {
    conferenceId,
    conference,
    speaker,
    video
  }
}

export const Video = compose<CombinedProps, Props>(
  connect(mapStateToProps),
  pure,
  withStateHandlers({ isOpen: false }, {
    toggleIsOpen: ({ isOpen }) => () => ({ isOpen: !isOpen })
  })
)(VideoInner)
