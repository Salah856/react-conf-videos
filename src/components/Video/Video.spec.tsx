import * as React from 'react'

import toJSON from 'enzyme-to-json'

import { mountWithStore, mockConference, mockVideo, wrapWithMemoryRouter } from 'utils'
import { VideoInner } from './Video'
import { Conference } from '../../domain'

describe('Video', () => {

  const getData = () => {
    const video = mockVideo()
    const conference: Conference = mockConference()
    const speaker = { name: 'Simon Carter' }
    const state = {
      data: {
        videos: {xxx: video},
        presenters: {aaa: speaker},
        conferences: {yyy : conference}
      }
    }
    const isOpen = false
    const props = { videoId: 'xxx', conferenceId: 'yyy', video, speaker, conference , isOpen, toggleIsOpen: jest.fn() }
    return { props, state }
  }

  it('should render and connect', () => {
    // arrange
    const { props, state } = getData()

    // act
    const wrapper = mountWithStore(state, wrapWithMemoryRouter(<VideoInner {...props} />))

    // assert
    expect(toJSON(wrapper)).toMatchSnapshot()
  })

  it('should toggle the isOpen state prop on click', () => {
    // arrange
    const { props, state } = getData()

    // act
    const wrapper = mountWithStore(state, wrapWithMemoryRouter(<VideoInner {...props} />))
    wrapper.find('.top').simulate('click')

    // assert
    expect(props.toggleIsOpen).toHaveBeenCalled()
  })

})
