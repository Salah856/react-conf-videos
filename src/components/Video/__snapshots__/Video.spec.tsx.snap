// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video should render and connect 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "xxxx",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "xxxx",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "xxxx",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Component
      conference={
        Object {
          "date": "XX/YY/ZZZ",
          "playlist": "day 1",
          "title": "fake conf title",
          "videos": Array [
            "aaa",
            "bbb",
            "ccc",
          ],
          "website": "fake url",
        }
      }
      conferenceId="yyy"
      isOpen={false}
      speaker={
        Object {
          "name": "Simon Carter",
        }
      }
      toggleIsOpen={[MockFunction]}
      video={
        Object {
          "embeddableLink": "a link",
          "length": "12:34",
          "link": "a link",
          "presenter": "aaa",
          "split": "",
          "title": "test title",
        }
      }
      videoId="xxx"
    >
      <div
        className="root"
        key="xxx"
      >
        <div
          className="top"
          onClick={[MockFunction]}
        >
          <span
            className="title"
          >
            test title
          </span>
          <span
            className="right"
          >
            12:34
          </span>
        </div>
        <div
          className="videoWrapper"
        />
        <div
          className="details"
        >
          <span>
            Simon Carter
          </span>
          <Link
            replace={false}
            to="/conference/yyy/fake-conf-title"
          >
            <a
              href="/conference/yyy/fake-conf-title"
              onClick={[Function]}
            >
              <span
                className="conferenceTitle"
              >
                fake conf title
              </span>
            </a>
          </Link>
        </div>
      </div>
    </Component>
  </Router>
</MemoryRouter>
`;
